include/GafferImage/DeepTileAccessor.h

```diff
@@ -0,0 +1,109 @@
+//        the distribution.
+//
+//      * Neither the name of Image Engine Design nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include "GafferImage/ImagePlug.h"
+#include "GafferImage/Sampler.h"
+
+#include <vector>
+
+namespace GafferImage
+{
+
+/// Utility class for sampling pixel values from a deep image.
+/// The interface mostly matches Sampler for flat images.
+class GAFFERIMAGE_API DeepTileAccessor
+{
+
+	public :
+
+		/// Sampler Constructor
+		/// @param plug The image plug to sample from.
+		/// @param channelName The channel to sample ( or empty string, if you only need sample counts )
+		/// @param sampleWindow The area from which samples may be requested. It is an error to request samples outside this area.
+		/// @param boundingMode The method of handling samples that fall outside the data window.
+		DeepTileAccessor( const GafferImage::ImagePlug *plug, const std::string &channelName, const Imath::Box2i &sampleWindow, Sampler::BoundingMode boundingMode = Sampler::Black );
+
+		/// Construct from another DeepTileAccessor with a different channelName, in order to reuse the
+		/// sample offsets data.
+		DeepTileAccessor( const DeepTileAccessor &source, const std::string &channelName );
+
+		/// Uses `parallelProcessTiles()` to fill the internal tile cache
+		/// with all tiles in the sample window. Allows `sample()` and
+		/// `visitPixels()` to subsequently be called concurrently.
+		void populate();
+
+		/// Gets the list of channel values and number of values at the specified
+		/// integer pixel coordinate. It is the caller's responsibility to ensure
+		/// that this point is contained within the sample window passed to the
+		/// constructor. If this was constructed with an empty channel name, data will
+		/// be set to nullptr
+		void sample( int x, int y, const float* &data, unsigned int &count );
```

Comment:
Sure, I'm happy to try out boost span and a separate sampleCount( x, y ). You're happy with the overall interface, with 'pass in channelName = "" if you just want counts"? I guess I can make sample() throw if you constructed it with an empty channel?

Reply:
> I guess I can make sample() throw if you constructed it with an empty channel?

Sounds good. Or it could just be UB, like when you pass a position which is outside the sample window.